Inductor cpp wrapper: fix dtype of ShapeAsConstantBuffer

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122297

For `at::scalar_tensor` the default dtype will be `float` ([link to scalar_tensor](https://github.com/pytorch/pytorch/blob/0d8e960f74acd359358e0b729c4803d2b71849e5/aten/src/ATen/native/TensorFactories.cpp#L856), [link to default dtype](https://github.com/pytorch/pytorch/blob/0d8e960f74acd359358e0b729c4803d2b71849e5/c10/core/TensorOptions.h#L551)) if we don't set the `dtype` value. However, the input scalar value is not necessarily a `float` value. With `torch::tensor(x)`, the dtype of the tensor will be decided according to the dtype of the scalar.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

chunyuan-w committed:
Inductor cpp wrapper: fix dtype of ShapeAsConstantBuffer

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix dtype of ShapeAsConstantBuffer"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on test/inductor/test_torchinductor.py:
```diff
@@ -2452,6 +2458,21 @@ def fn(x, y):
+            return (full, arg1_1, embedding)
+
+        arg0_1 = rand_strided((32128, 768), (768, 1), device="cpu", dtype=torch.float32)
+        arg1_1 = 22
```
`arg1_1` is not used in this scope, delete it?

chunyuan-w replied:
Deleted the unused variable

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix dtype of ShapeAsConstantBuffer"


For `at::scalar_tensor` the default dtype will be `float` ([link to scalar_tensor](https://github.com/pytorch/pytorch/blob/0d8e960f74acd359358e0b729c4803d2b71849e5/aten/src/ATen/native/TensorFactories.cpp#L856), [link to default dtype](https://github.com/pytorch/pytorch/blob/0d8e960f74acd359358e0b729c4803d2b71849e5/c10/core/TensorOptions.h#L551)) if we don't set the `dtype` value. However, the input scalar value is not necessarily a `float` value. With `torch::tensor(x)`, the dtype of the tensor will be decided according to the dtype of the scalar.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

chunyuan-w commented:
@desertfire I just noticed that https://github.com/pytorch/pytorch/pull/118024 changed `torch::tensor` to `at::scalar_tensor`. May I know if there's a specific reason for this? I found that with `at::scalar_tensor`, the scalar type of the tensor might be inconsistent with the scalar value and will cause regression issue of several models in https://github.com/pytorch/pytorch/issues/122292.

chunyuan-w commented:
> > @desertfire I just noticed that #118024 changed `torch::tensor` to `at::scalar_tensor`. May I know if there's a specific reason for this? I found that with `at::scalar_tensor`, the scalar type of the tensor might be inconsistent with the scalar value and will cause regression issue of several models in #122292.
> 
> It is related to https://github.com/pytorch/pytorch/pull/118024/files/bba213c151f2c8e7a29635d273a73a6c98d24393#r1463364631. `at::scalar_tensor` is more accurate in this use case. Is it a dtype propagation problem here?

Considering an int scalar value `2`, `at::scalar_tensor(2)` becomes a Float type tensor.
The issues in https://github.com/pytorch/pytorch/issues/122292 is that this scalar value output is later used as a `size` which requires it to be an int instead of a float.

I tried to get the dtype of the scalar value and pass it to `at::scalar_tensor` as the `dtype` input argument during the codegen time but sometimes the scalar value is a `Symbol` and I failed to find a good way to infer the dtype of this `Symbol`, while with `torch::tensor`, the dtype of the output tensor will match the dtype of the scalar input value, so I changed `at::scalar_tensor` back to `torch::tensor` in non abi compatible mode. May I know if you have other suggestions for the fix of this issue?

chunyuan-w commented:
Hi @desertfire this issue will cause regression against the PyTorch 2.2 release and we're trying to see if it's possible to fix it in 2.3. May I know if you have other suggestions regarding the current fix approach?

desertfire reviewed (approved):
ok for the non abi-compatible mode, although we will have to come back to fix once I turned on the abi-compatible mode as default.

chunyuan-w commented:
@pytorchbot merge

chunyuan-w commented:
> ok for the non abi-compatible mode, although we will have to come back to fix once I turned on the abi-compatible mode as default.

Oh okay. Btw, when do we plan to turn on the abi-compatible mode as default?

desertfire commented:
> > ok for the non abi-compatible mode, although we will have to come back to fix once I turned on the abi-compatible mode as default.
> 
> Oh okay. Btw, when do we plan to turn on the abi-compatible mode as default?

We do, once its coverage has reached a reasonable level.
